ts/routes/image-occlusion/tools/lib.ts

```diff
@@ -86,8 +93,12 @@ export const unGroupShapes = (canvas: fabric.Canvas): void => {
     group.destroyed = true;
     canvas.remove(group);
 
-    items.forEach((item) => {
-        item.set({ opacity: get(opacityStateStore) ? 0.4 : 1 });
+    // @ts-expect-error not defined
+    const maxOrdinal = Math.max(canvas.getObjects().map((item) => item.ordinal));
+
+    items.forEach((item, index) => {
+        // @ts-expect-error not defined
+        item.set({ opacity: get(opacityStateStore) ? 0.4 : 1, ordinal: maxOrdinal + index + 1 });
```

Comment:
This is not working as expected. It can result in two unrelated masks getting the same ordinal, causing them to be grouped. To reproduce:
1. Add 3 masks
2. Group two of the masks
3. Ungroup
4. Reload the note (e.g. by selecting another browser note then back to the note)
5. Notice one mask from the original group is grouped with the third.

Reply:
I have updated the implementation, it should work now, but order cannot be maintained in case of ungrouping of shapes.

Reply:
No problem, we can go with this.

Reply:
@dae While the issue I reported is fixed, this is re-introducing the original problem fixed by #2987 when there's an ungroup operation. Unrelated masks could get reordered.

This appears to work better:
```diff
diff --git a/ts/routes/image-occlusion/shapes/to-cloze.ts b/ts/routes/image-occlusion/shapes/to-cloze.ts
index 01d8b62c9..fbdd03f07 100644
--- a/ts/routes/image-occlusion/shapes/to-cloze.ts
+++ b/ts/routes/image-occlusion/shapes/to-cloze.ts
@@ -3,9 +3,7 @@

 import { fabric } from "fabric";
 import { cloneDeep } from "lodash-es";
-import { get } from "svelte/store";

-import { shapeUngroupState } from "../store";
 import { getBoundingBox } from "../tools/lib";
 import type { Size } from "../types";
 import type { Shape, ShapeOrShapes } from "./base";
@@ -22,6 +20,7 @@ export function exportShapesToClozeDeletions(occludeInactive: boolean): {

     let clozes = "";
     let index = 0;
+    let nextOrdinal = 1;
     shapes.forEach((shapeOrShapes) => {
         // shapes with width or height less than 5 are not valid
         if (shapeOrShapes === null) {
@@ -31,7 +30,33 @@ export function exportShapesToClozeDeletions(occludeInactive: boolean): {
         if (shapeOrShapes instanceof Rectangle && shapeOrShapes.fill === "transparent") {
             return;
         }
-        clozes += shapeOrShapesToCloze(shapeOrShapes, index, occludeInactive);
+        // Maintain existing ordinal in editing mode
+        let ordinal: number | undefined;
+        if (Array.isArray(shapeOrShapes)) {
+            ordinal = shapeOrShapes[0].ordinal;
+        } else {
+            ordinal = shapeOrShapes.ordinal;
+        }
+        if (ordinal === undefined) {
+            if (shapeOrShapes instanceof Text) {
+                ordinal = 0;
+            } else {
+                ordinal = nextOrdinal;
+            }
+            if (Array.isArray(shapeOrShapes)) {
+                shapeOrShapes.forEach((shape) => (shape.ordinal = ordinal));
+            } else {
+                shapeOrShapes.ordinal = ordinal;
+            }
+        }
+        nextOrdinal = ordinal + 1;
+
+        clozes += shapeOrShapesToCloze(
+            shapeOrShapes,
+            index,
+            ordinal,
+            occludeInactive,
+        );
         if (!(shapeOrShapes instanceof Text)) {
             index++;
         }
@@ -134,6 +159,7 @@ function fabricObjectToBaseShapeOrShapes(
 function shapeOrShapesToCloze(
     shapeOrShapes: ShapeOrShapes,
     index: number,
+    ordinal: number,
     occludeInactive: boolean,
 ): string {
     let text = "";
@@ -145,7 +171,7 @@ function shapeOrShapesToCloze(
     let type: string;
     if (Array.isArray(shapeOrShapes)) {
         return shapeOrShapes
-            .map((shape) => shapeOrShapesToCloze(shape, index, occludeInactive))
+            .map((shape) => shapeOrShapesToCloze(shape, index, ordinal, occludeInactive))
             .join("");
     } else if (shapeOrShapes instanceof Rectangle) {
         type = "rect";
@@ -166,16 +192,6 @@ function shapeOrShapesToCloze(
         addKeyValue("oi", "1");
     }

-    // Maintain existing ordinal in editing mode
-    let ordinal = shapeOrShapes.ordinal;
-    if (ordinal === undefined || Number.isNaN(ordinal) || get(shapeUngroupState)) {
-        if (type === "text") {
-            ordinal = 0;
-        } else {
-            ordinal = index + 1;
-        }
-        shapeOrShapes.ordinal = ordinal;
-    }
     text = `{{c${ordinal}::image-occlusion:${type}${text}}}<br>`;

     return text;
diff --git a/ts/routes/image-occlusion/store.ts b/ts/routes/image-occlusion/store.ts
index b1653f4fa..5847250dd 100644
--- a/ts/routes/image-occlusion/store.ts
+++ b/ts/routes/image-occlusion/store.ts
@@ -17,5 +17,3 @@ export const ioImageLoadedStore = writable(false);
 export const opacityStateStore = writable(false);
 // store state of text editing
 export const textEditingState = writable(false);
-// store state of shape ungroup
-export const shapeUngroupState = writable(false);
diff --git a/ts/routes/image-occlusion/tools/lib.ts b/ts/routes/image-occlusion/tools/lib.ts
index 9a6738b10..bc0606cdc 100644
--- a/ts/routes/image-occlusion/tools/lib.ts
+++ b/ts/routes/image-occlusion/tools/lib.ts
@@ -4,7 +4,7 @@
 import { fabric } from "fabric";
 import { get } from "svelte/store";

-import { opacityStateStore, shapeUngroupState } from "../store";
+import { opacityStateStore } from "../store";

 export const SHAPE_MASK_COLOR = "#ffeba2";
 export const BORDER_COLOR = "#212121";
@@ -94,11 +94,14 @@ export const unGroupShapes = (canvas: fabric.Canvas): void => {
     canvas.remove(group);

     items.forEach((item) => {
-        item.set({ opacity: get(opacityStateStore) ? 0.4 : 1 });
+        item.set({
+            opacity: get(opacityStateStore) ? 0.4 : 1,
+            // @ts-expect-error not defined
+            ordinal: undefined,
+        });
         canvas.add(item);
     });

-    shapeUngroupState.set(true);
     redraw(canvas);
 };

```

What do you think @dae @krmanik ?

Reply:
This does not seem to work, create more 4 shapes, group shape, so group 1 has two shapes and group 2 has two shapes, click on add button, the error is (You have a cloze notetype but have not made any cloze deletions. Proceed?)
I think we have spent enough time solving this, we should continue without thinking about order as it is not breaking changes.

This updated patch based on above on latest main commit,

<details>
<summary>Code</summary>

```diff
diff --git a/ts/routes/image-occlusion/shapes/to-cloze.ts b/ts/routes/image-occlusion/shapes/to-cloze.ts
index 01d8b62c9..fbdd03f07 100644
--- a/ts/routes/image-occlusion/shapes/to-cloze.ts
+++ b/ts/routes/image-occlusion/shapes/to-cloze.ts
@@ -3,9 +3,7 @@
 
 import { fabric } from "fabric";
 import { cloneDeep } from "lodash-es";
-import { get } from "svelte/store";
 
-import { shapeUngroupState } from "../store";
 import { getBoundingBox } from "../tools/lib";
 import type { Size } from "../types";
 import type { Shape, ShapeOrShapes } from "./base";
@@ -22,6 +20,7 @@ export function exportShapesToClozeDeletions(occludeInactive: boolean): {
 
     let clozes = "";
     let index = 0;
+    let nextOrdinal = 1;
     shapes.forEach((shapeOrShapes) => {
         // shapes with width or height less than 5 are not valid
         if (shapeOrShapes === null) {
@@ -31,7 +30,33 @@ export function exportShapesToClozeDeletions(occludeInactive: boolean): {
         if (shapeOrShapes instanceof Rectangle && shapeOrShapes.fill === "transparent") {
             return;
         }
-        clozes += shapeOrShapesToCloze(shapeOrShapes, index, occludeInactive);
+        // Maintain existing ordinal in editing mode
+        let ordinal: number | undefined;
+        if (Array.isArray(shapeOrShapes)) {
+            ordinal = shapeOrShapes[0].ordinal;
+        } else {
+            ordinal = shapeOrShapes.ordinal;
+        }
+        if (ordinal === undefined) {
+            if (shapeOrShapes instanceof Text) {
+                ordinal = 0;
+            } else {
+                ordinal = nextOrdinal;
+            }
+            if (Array.isArray(shapeOrShapes)) {
+                shapeOrShapes.forEach((shape) => (shape.ordinal = ordinal));
+            } else {
+                shapeOrShapes.ordinal = ordinal;
+            }
+        }
+        nextOrdinal = ordinal + 1;
+
+        clozes += shapeOrShapesToCloze(
+            shapeOrShapes,
+            index,
+            ordinal,
+            occludeInactive,
+        );
         if (!(shapeOrShapes instanceof Text)) {
             index++;
         }
@@ -134,6 +159,7 @@ function fabricObjectToBaseShapeOrShapes(
 function shapeOrShapesToCloze(
     shapeOrShapes: ShapeOrShapes,
     index: number,
+    ordinal: number,
     occludeInactive: boolean,
 ): string {
     let text = "";
@@ -145,7 +171,7 @@ function shapeOrShapesToCloze(
     let type: string;
     if (Array.isArray(shapeOrShapes)) {
         return shapeOrShapes
-            .map((shape) => shapeOrShapesToCloze(shape, index, occludeInactive))
+            .map((shape) => shapeOrShapesToCloze(shape, index, ordinal, occludeInactive))
             .join("");
     } else if (shapeOrShapes instanceof Rectangle) {
         type = "rect";
@@ -166,16 +192,6 @@ function shapeOrShapesToCloze(
         addKeyValue("oi", "1");
     }
 
-    // Maintain existing ordinal in editing mode
-    let ordinal = shapeOrShapes.ordinal;
-    if (ordinal === undefined || Number.isNaN(ordinal) || get(shapeUngroupState)) {
-        if (type === "text") {
-            ordinal = 0;
-        } else {
-            ordinal = index + 1;
-        }
-        shapeOrShapes.ordinal = ordinal;
-    }
     text = `{{c${ordinal}::image-occlusion:${type}${text}}}<br>`;
 
     return text;
diff --git a/ts/routes/image-occlusion/store.ts b/ts/routes/image-occlusion/store.ts
index b1653f4fa..5847250dd 100644
--- a/ts/routes/image-occlusion/store.ts
+++ b/ts/routes/image-occlusion/store.ts
@@ -17,5 +17,3 @@ export const ioImageLoadedStore = writable(false);
 export const opacityStateStore = writable(false);
 // store state of text editing
 export const textEditingState = writable(false);
-// store state of shape ungroup
-export const shapeUngroupState = writable(false);
diff --git a/ts/routes/image-occlusion/tools/lib.ts b/ts/routes/image-occlusion/tools/lib.ts
index 9a6738b10..ea60ad3d3 100644
--- a/ts/routes/image-occlusion/tools/lib.ts
+++ b/ts/routes/image-occlusion/tools/lib.ts
@@ -4,7 +4,7 @@
 import { fabric } from "fabric";
 import { get } from "svelte/store";
 
-import { opacityStateStore, shapeUngroupState } from "../store";
+import { opacityStateStore } from "../store";
 
 export const SHAPE_MASK_COLOR = "#ffeba2";
 export const BORDER_COLOR = "#212121";
@@ -94,11 +94,14 @@ export const unGroupShapes = (canvas: fabric.Canvas): void => {
     canvas.remove(group);
 
     items.forEach((item) => {
-        item.set({ opacity: get(opacityStateStore) ? 0.4 : 1 });
+        item.set({
+            opacity: get(opacityStateStore) ? 0.4 : 1,
+            // @ts-expect-error not defined
+            ordinal: undefined,
+         });
         canvas.add(item);
     });
 
-    shapeUngroupState.set(true);
     redraw(canvas);
 };
 

```

</details>

Reply:
Nice catch @abdnh! And thanks for the quick fix @krmanik.

Unfortunately I ran into another issue. I created three rectangles and merged them into a group, then created three more, and left them ungrouped. I added the note. Then I edited it, and ungrouped the three rectangles. Two surprises:

- The group's ordinal was not reused, so the first card is now empty and needs to be cleaned up with tools>empty cards
- One of the shapes in the group was given the same ordinal as one of the shapes that was never in a group, so it ended up in a new group.